### ISSUE TITLE
Windows shim not working when on WSL directory

### ISSUE THREAD

**Issue by felipecrs** [volta-cli/volta#966]
I'm on Windows, but accessing a folder within my WSL distro.

```sh-session
➜ npm --version
'\\wsl$\Ubuntu\home\felipecrs\repos\banana'
CMD.EXE was started with the above path as the current directory.
UNC paths are not supported.  Defaulting to Windows directory.
7.5.2
```

This causes `npm install` not to work, as it thinks it is on `C:\Windows`.

**Comment by charlespierce** [comment]
@felipecrs Thanks for reporting! Unfortunately, Volta makes significant use of `cmd.exe` internally to handle executing the various commands. It seems that `cmd.exe` itself doesn't support the path used for accessing the WSL files from Windows proper, so I'm not sure if there's a clear solution.

The workaround would be to install Volta in the WSL environment (using the [Unix instructions](https://docs.volta.sh/guide/getting-started)) and then run `npm install` from there.

**Comment by felipecrs** [comment]
Oh yeah, I use volta inside WSL as well. But I needed to run it in Windows because I was testing a UI app. I'll investigate on how this can be solved anyway.

**Comment by charlespierce** [comment]
One thought that may work (I don't know how deep the restrictions on UNC paths go): You could run `volta which npm` from that directory to determine the `npm` file that would actually be executed, and then attempt to call that directly. That said, I would advocate caution when running a command like `npm install` from Windows inside of the WSL directory, as it may result in native modules being compiles for Windows and then breaking things when you go back to the WSL side.

Another solution, though more annoying to coordinate, would be to clone the repo to your normal Windows home directory, and then work within that separate environment.

**Comment by felipecrs** [comment]
> That said, I would advocate caution when running a command like npm install from Windows inside of the WSL directory, as it may result in native modules being compiles for Windows and then breaking things when you go back to the WSL side.

For sure, I'm aware of the complications.

> Another solution, though more annoying to coordinate, would be to clone the repo to your normal Windows home directory, and then work within that separate environment.

Yeah, it would also work by moving the contents from WSL to Windows.

I'm not trying to circumvent my need, I did it already by using the `npm` installed with Chocolatey, but I'll take a look at your solution with `volta which npm`.

I just wanted to report this issue as none reported earlier, and if we solve we can make volta better. :D

**Comment by charlespierce** [comment]
> But anyway, I'm not trying to circumvent my need... I just wanted to report this issue as none was reported earlier, and if we solve, we can make volta better. :D

Makes sense, thanks!

**Comment by felipecrs** [comment]
Wow, ok, it doesn't seem to be a volta issue at all:

```sh-session
➜ Get-Command npm

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Application     npm.cmd                                            0.0.0.0    C:\Program Files\nodejs\npm.cmd
➜ npm --version
'\\wsl$\Ubuntu\home\felipecrs\repos\banana'
CMD.EXE was started with the above path as the current directory.
UNC paths are not supported.  Defaulting to Windows directory.
7.7.6
```

**Comment by charlespierce** [comment]
Yeah, that was one of my concerns: It looks like the scripts that `npm` itself uses to launch the JS file through Node rely on `cmd.exe`, so since that doesn't play well with the UNC paths I'm not sure if it'll work directly. You may be able to work even further around it by running `C:\path\to\node.exe <path to npm-cli.js> <arguments>`, but even then you might run into issues with what `npm` does internally, I'm not sure.

**Comment by felipecrs** [comment]
Ok, just tested some things here and:

1. All the volta shims are broken (as they use cmd.exe)
2. Npm's own shim is also broken (as it also uses cmd.exe): https://github.com/npm/cli/issues/1189

**Comment by felipecrs** [comment]
Yeah... I think the only possible solution would be to use PowerShell for the shim instead of cmd. But it's not portable to Windows 7.

Would it be possible to create the shims using Rust itself instead?

**Comment by charlespierce** [comment]
> Would it be possible to create the shims using Rust itself instead?

The core of the shims is already in Rust, `cmd` is used when passing control to the underlying command (e.g. `npm` or `node`). Windows 7 support isn't a major concern, as I don't believe Volta's symlink approach works under Windows 7 anyway. However, I'm not sure if Powershell would execute the commands in the same way. It's been a while since I messed around with all the misdirection in Windows, but I know there are a lot of edge cases to make sure that it works in the normal command prompt, powershell, _and_ emulation layers like Git Bash / MinGW.

We do have an issue already around looking to replace the symlinks with a lightweight tool #579

**Comment by felipecrs** [comment]
> cmd is used when passing control to the underlying command (e.g. npm or node)

Hm... got it. I wonder if can't `npm` or `node` be called without `cmd`. On Linux it's possible to call binaries without a shell at all, but I don't know how it would work on Windows.

**Comment by charlespierce** [comment]
They certainly _can_ be called without using `cmd.exe`, as Volta running at all means that there is a shell executing it. The difficulty is that each different shell has very different semantics for how spawning a new process works. It’s also not always trivial to determine which shell is running, so smoothing over those inconsistencies can be tricky.

That said, it’s definitely worth taking another look to see if there’s a cleaner approach.

**Comment by felipecrs** [comment]
It may be a silly question, but I really did not understand: why is `cmd.exe` or any shell present in the loop?

In my head, it could be like the following:

`C:\volta\bin\node.exe` (Rust compiled binary) -> `C:\volta\[...]\node.exe` (Original node binary)

Or, for NPM:

`C:\volta\bin\npm.exe` (Rust compiled binary) -> `C:\volta\[...]\node.exe path\to\npm-cli.js [args]` 

So, I don't know where `cmd.exe` fits in this flow.

**Comment by charlespierce** [comment]
There's a few reasons: First is that when handling an arbitrary package, it's not clear whether the file we're trying to run when executing `npm` is an EXE, CMD, or BAT file. `cmd.exe` handles that resolution for us. Additionally, I believe `cmd.exe` is needed to execute BAT or CMD files anyway (and powershell calls that under the hood, I think).

Next is that Rust's `std::process::Command` has some strange semantics when changing the `Path` environment variable while creating a new process. That is how Volta works internally, so using `cmd.exe /C <command>` helps us avoid infinite recursion (where the `node` shim finds itself over and over).

I think there may be some other issues, I'm not 100% as it's been a little while since I was deep in the weeds there. We could potentially work around them with our own implementation that handles the resolution and the various PATH issues. However, that would likely be more brittle and prone to breaking than using the OS-provided tools. Similarly, we might (in some cases) be able to detect that the file is a JS file and execute it using `node <path to script>`, but that would mean we are duplicating the behavior of `npm` itself and again making things less future-proof.

**Comment by felipecrs** [comment]
Thank you so much for the elaborated answer.

> Similarly, we might (in some cases) be able to detect that the file is a JS file and execute it using `node <path to script>`, but that would mean we are duplicating the behavior of `npm` itself and again making things less future-proof.

I don't think we should go in that route unless it's really needed. I used `npm` as example, but in fact even the Volta shim for `node.exe` is having the same issue. So, I guess we can leave the `npm`'s own problem (builtin cmd.exe shims which call node.exe to call the js file) out of the scope here.

**Comment by felipecrs** [comment]
As explained in https://github.com/npm/cli/issues/1189#issuecomment-817209410, https://github.com/kiennq/scoop-better-shimexe/pull/1 seems to be a viable solution to this issue.
